controllers/extendeddaemonsetreplicaset/strategy/canary.go

```diff
@@ -29,6 +36,24 @@ func ManageCanaryDeployment(client client.Client, daemonset *v1alpha1.ExtendedDa
 	var err error
 	isPaused, _ := eds.IsCanaryDeploymentPaused(daemonset.GetAnnotations())
 
+	var autoPauseEnabled bool
+	var maxRestarts int
+	if daemonset.Spec.Strategy.Canary == nil || daemonset.Spec.Strategy.Canary.AutoPause == nil {
+		autoPauseEnabled = defaultAutoPauseEnabled
+		maxRestarts = defaultMaxRestarts
+	} else {
+		if daemonset.Spec.Strategy.Canary.AutoPause.Enabled == nil {
+			autoPauseEnabled = defaultAutoPauseEnabled
+		} else {
+			autoPauseEnabled = *daemonset.Spec.Strategy.Canary.AutoPause.Enabled
+		}
+		if daemonset.Spec.Strategy.Canary.AutoPause.MaxRestarts == nil {
+			maxRestarts = defaultMaxRestarts
+		} else {
+			maxRestarts = int(*daemonset.Spec.Strategy.Canary.AutoPause.MaxRestarts)
+		}
+	}
```

Comment:
please correct me if I'm wrong, but I think at this point the defaulting should be already done via `DefaultExtendedDaemonSetSpecStrategyCanaryAutoPause` ?

Reply:
to be clearer, the question is: why do we default the autopause config twice?

Reply:
Thanks for bringing this up, as I didn't have a complete grasp on this.

It seems we `default` the ERS (in extendeddaemonset_default.go) if the minimum required settings are not defined. In this case we could set the default AutoPause settings with `DefaultExtendedDaemonSetSpecStrategyCanaryAutoPause()`, but it's not strictly necessary since we also have this defaulting code here in canary.go. If you think it's confusing to include the change in extendeddaemonset_default.go I can remove it.

Reply:
thanks for clarifying this
tbh I would expect the defaulting to be done in one place (in `extendeddaemonset_default.go`), the fact that the same defaulting logic and the same constants (`defaultMaxRestarts` `defaultAutoPauseEnabled` `defaultCanaryAutoPauseEnabled` `defaultCanaryAutoPauseMaxRestarts`) are defined in different packages doesn't seem right 🤔

Reply:
I think if we do it as @ahmed-mez  said in `extendeddaemonset_default.go`, the defaulting is not needed in `ManageCanaryDeployment` because we always call the default method.
To achieve it, the function  `IsDefaultedExtendedDaemonSet()` [link](https://github.com/DataDog/extendeddaemonset/blob/master/api/v1alpha1/extendeddaemonset_default.go#L25), should be updated, to return false if the canary configuration is missing default values.

Reply:
I tested it like this, with only setting the defaults in extendeddaemonset_default.go, and it led to undesired behaviors. I will try again.

Reply:
OK got this fixed, and confirmed it works whether `autoPause`, `autoPause.Enabled`, or `autoPause.MaxRestarts` is missing. Sorry for the confusion